sklearn/metrics/_regression.py

```diff
@@ -1177,8 +1202,34 @@ def r2_score(
     >>> r2_score(y_true, y_pred, force_finite=False)
     -inf
     """
-    y_type, y_true, y_pred, multioutput = _check_reg_targets(
-        y_true, y_pred, multioutput
+    input_arrays = [y_true, y_pred]
+    if sample_weight is not None:
+        input_arrays.append(sample_weight)
+
+    multioutput_is_array = multioutput is not None and not isinstance(multioutput, str)
+    if multioutput_is_array:
+        input_arrays.append(multioutput)
+
+    xp, is_array_api_compliant = get_namespace(*input_arrays)
+    input_xp = xp
+    device_ = device(*input_arrays)
+
+    if not _supports_dtype(xp, device_, "float64"):
+        y_true = _convert_to_numpy(y_true, xp)
+        y_pred = _convert_to_numpy(y_pred, xp)
+        if sample_weight is not None:
+            sample_weight = _convert_to_numpy(sample_weight, xp)
+        if multioutput_is_array:
+            multioutput = _convert_to_numpy(multioutput, xp)
+        xp, _ = get_namespace(y_true)
+        device_ = device(y_true)
+
+    dtype = (
+        "numeric" if not is_array_api_compliant else supported_float_dtypes(xp, device_)
+    )
+
+    _, y_true, y_pred, multioutput = _check_reg_targets(
+        y_true, y_pred, multioutput, dtype=dtype, xp=xp
```

Comment:
Trying to read this diff is quite hard for somebody like me who hasn't done too much array API work. I think this would contribute to maintenance burden quite a bit. How can we simplify this in a way that it's easier to understand and maintain, and maybe a shorter diff?

Reply:
I think this change is the in `r2_score` PR already, so unrelated to "fixing `train_test_split`". Maybe worth raising in https://github.com/scikit-learn/scikit-learn/pull/27904